flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -571,9 +525,11 @@ private void runJob(JobGraph jobGraph, ExecutionType executionType) throws Excep
                                 getMainThreadExecutor());
 
         final CompletableFuture<Void> jobTerminationFuture =
-                cleanupJobStateFuture
-                        .thenApply(cleanupJobState -> removeJob(jobId, cleanupJobState))
-                        .thenCompose(Function.identity());
+                cleanupJobStateFuture.thenCompose(
+                        cleanupJobState -> {
+                            getMainThreadExecutor().assertRunningInMainThread();
```

Comment:
true, I also remove it from `Dispatcher.globalCleanup(JobID)` where it's also not really necessary. I will leave it in for `Dispatcher.localCleanup(JobID)`, though, because it's a requirement for the `DefaultResourceCleaner.cleanupAsync(JobID)` and having that assert makes it more explicit.